### PR TITLE
ADD: Docker support to example project

### PR DESCRIPTION
**Pull request by jamsea**
Currently working on the latest version of docker for mac. The only gross part atm is you need a set an environment variable on your host called `JSPM_GITHUB_AUTH_TOKEN` in order for JSPM to work inside the container.

Still adding more changes, planning to add all of the wildcat environment variables

### CODE REVIEW

**Comment by jamsea** [comment]
File watching between the host OS and the container is working fine, oddly file watching from the app server inside the container isn't working.

File: example/Dockerfile [hunk 1]
```diff
@@ -0,0 +1,17 @@
+FROM node:4.4.4
+ENV PATH ./node_modules/.bin/:$PATH
+ENV NPM_CONFIG_LOGLEVEL warn
```
Comment (jamsea) [inline example/Dockerfile]: By default the official node image sets this to verbose, which produces way too many logs. Set it back to warn (default for npm when you install it on os x)

File: package.json [hunk 2]
```diff
@@ -35,7 +35,7 @@
     "estraverse-fb": "^1.3.1",
     "isomorphic-fetch": "^2.2.1",
     "istanbul": "https://github.com/nfl/istanbul/tarball/6a3222a5bca6e38af456c1c18d724e1daf69533d",
-    "jspm": "nfl/jspm-cli#0.16.34/fetch-support",
+    "jspm": "github:nfl/jspm-cli#0.16.34/fetch-support",
```
Comment (doctyper) [inline package.json]: ? Please revert this.

**Comment by doctyper** [comment]
@jamsea Gonna need to bump the coverage up on this.

**Comment by jamsea** [comment]
Got it working. `docker-compose up -d` will run `npm run dev` by default. To run the e2e tests type `docker-compose run test:e2e`.

To get the dns resolving working on the host machine create the file `/etc/resolver/dev` that contains the following:
```
nameserver 127.0.0.1
port 53535
```

File: example/.eslintrc [hunk 3]
```diff
@@ -13,6 +13,9 @@
         "expect": true,
         "protractor": true
     },
+    "rules": {
+        "prefer-arrow-callback": 0
+    },
```
Comment (doctyper) [inline example/.eslintrc]: Not sure why this is being added?

File: example/Dockerfile [hunk 4]
```diff
@@ -0,0 +1,22 @@
+FROM node:4.4.4
```
Comment (doctyper) [inline example/Dockerfile]: Consider bumping to 6.6.0 to retain parity with our production environment. If you do, don't forget to add a test for 6.6.0 in `.travis.yml`

File: example/Dockerfile [hunk 5]
```diff
@@ -0,0 +1,22 @@
...
+# Github auth token has to be passed in during build time, unfortunately
+# the only way to do this is via an environment variable on the host
+# machine that must be set. To generate a token use `jspm registry config github`
+# This is the only dependency outside of docker that's needed
+ARG JSPM_GITHUB_AUTH_TOKEN
+
+RUN mkdir /code
+WORKDIR /code
+
+RUN npm install -g npm@3.10.5
```
Comment (doctyper) [inline example/Dockerfile]: Should check if this is the latest npm version.

File: README.md [hunk 6]
```diff
@@ -60,27 +60,49 @@ For first-time jspm users, you will need to configure it to use GitHub:
 
 Follow the steps to enable GitHub access.
 
-### Getting started
+### Getting started: docker development
+Clone this repository.
+
+Create the file `/etc/resolver/dev` with the following contents:
```
Comment (jamsea) [inline README.md]: Yup, it's an absolute must to get DNS resolving between the host and the container working on OS X, and it does need root permissions to accomplish (similar to editing /etc/hosts)

Comment (doctyper) [inline README.md]: Can you add a note that this might/will require permissions?

File: example/.eslintrc (hunk 3 above)
Comment (jamsea) [inline example/.eslintrc]: I'm getting ~100 `error  Unexpected function expression  prefer-arrow-callback` errors with eslint in wildcat. I'm guessing something somewhere has changed. I know for this project we prefer named function callbacks

Comment (doctyper) [inline example/.eslintrc]: That should only report if it finds an anonymous function. Named functions should be fine. My guess is it found an anon function it doesn't like.

Comment (jamsea) [inline example/.eslintrc]: Nevermind, fix was already in master I just haven't merged the changes in

Comment (doctyper) [inline example/.eslintrc]: Cool, so can you revert this?

Comment (jamsea) [inline example/.eslintrc]: Just pushed all the changes to all the comments a second ago 👍  They don't appear in this view though, seems like you have to click on the "Files" tab in Github to see them

**Comment by doctyper** [comment]
This is unfortunately still failing testing via `make install`. I think it's the invalid localhost certificates. I'm also seeing a Webpack warning, so the config needs to be updated:

```
Invalid configuration object. Webpack has been initialised using a configuration object that does not match the API schema.
 - configuration has an unknown property 'debug'. These properties are valid:
   object { amd?, bail?, cache?, context?, dependencies?, devServer?, devtool?, entry, externals?, loader?, module?, name?, node?, output?, plugins?, profile?, recordsInputPath?, recordsOutputPath?, recordsPath?, resolve?, resolveLoader?, stats?, target?, watch?, watchOptions? }
   The 'debug' property was removed in webpack 2.
   Loaders should be updated to allow passing this option via loader options in module.rules.
   Until loaders are updated one can use the LoaderOptionsPlugin to switch loaders into debug mode:
   plugins: {
     new webpack.LoaderOptionsPlugin({
       debug: true
     })
   }
 - configuration.resolve has an unknown property 'modulesDirectories'. These properties are valid:
   object { alias?, aliasFields?, cachePredicate?, descriptionFiles?, enforceExtension?, enforceModuleExtension?, extensions?, fileSystem?, mainFields?, mainFiles?, moduleExtensions?, modules?, plugins?, resolver?, symlinks?, unsafeCache? }
```

File: example/protractor.config.js [hunk 7]
```diff
@@ -39,6 +39,7 @@ const e2eReportDir = coverageSettings.e2e.reporting.dir;
 
 /* global browser */
 exports.config = {
+    seleniumAddress: appServerSettings.hostname === "localhost" ? null : "http://selenium:4444/wd/hub",
```
Comment (doctyper) [inline example/protractor.config.js]: This selenium address should probably come from wildcat.config.js

File: example/wildcat.config.js [hunk 8]
```diff
@@ -1,7 +1,41 @@
...
+
 const pkg = require("./package.json");
 const __DEV__ = (process.env.NODE_ENV === "development");
 const __PROD__ = (process.env.NODE_ENV === "production");
 const __BUNDLE__ = process.env.COVERAGE !== "e2e";
+const __IN_DOCKER__ = fs.existsSync("/.dockerenv");
+
+
+function getDefaultSSLFile(filename) {
+    const filePath = __IN_DOCKER__ ? "ssl/example." : "../packages/react-wildcat/ssl/server.";
```
Comment (doctyper) [inline example/wildcat.config.js]: This seems very fragile. What if I have a `.dockerenv` file but also run `npm run dev` outside of Docker? This file has access to your current hostname, you should be able to set up certificates based on pathname. e.g.:

```
- ssl
  - example.dev
    - server.key
    - server.cert
    - server.ca
  - localhost
    - server.key
    - server.cert
    - server.ca
```

File: shell/install.js [hunk 9]
```diff
@@ -17,6 +17,11 @@ ls("packages").forEach((loc) => {
     const pkg = require(pkgPath);
 
     cd(pkgDir);
+
+    if (pkgDir === "react-wildcat-handoff") {
+        exec(`npm run build`);
```
Comment (doctyper) [inline shell/install.js]: Can you explain this add? You shouldn't need to manually do this.

Comment (jamsea) [inline shell/install.js]: This has to get built _before_ jspm linking happens, or else you'll receive a bunch of 404 errors on react-wildcat-handoff/dist/client.js

Comment (jamsea) [inline shell/install.js]: This is actually broken in `master` right now too

File: example/wildcat.config.js (hunk 8 above)
Comment (doctyper) [inline example/wildcat.config.js]: You're attempting to reference `wildcatConfig` before it's defined.

File: .node-version [hunk 10]
```diff
@@ -1 +1 @@
-4.4.4
+6.6.0
```
Comment (doctyper) [inline .node-version]: I tried reverting to 4.4.0 and the 404 error does not occur, so this is either a long-standing issue surfaced by Node 6, or a Node 6 specific issue.

Comment (doctyper) [inline .node-version]: Update: Did not see the error using Node 6.3.1.

Comment (jamsea) [inline .node-version]: Cool I just tried it with 4.4.4 and it was fine too. I'll bump it to 6.3.1. Really odd that the build step is consistently not working for handoff in 6.6.0 when you run `make install`

File: example/wildcat.config.js [hunk 11]
```diff
@@ -1,8 +1,41 @@
...
+const path = require("path");
+
 const pkg = require("./package.json");
 const __DEV__ = (process.env.NODE_ENV === "development");
 const __PROD__ = (process.env.NODE_ENV === "production");
 const __BUNDLE__ = process.env.COVERAGE !== "e2e";
 
+function getDefaultSSLFile(filename) {
+    const filePath = process.env.HOST === "localhost" || !process.env.HOST ? "../packages/react-wildcat/ssl/server." : "ssl/example.";
+    console.log(process.env.HOST);
```
Comment (doctyper) [inline example/wildcat.config.js]: Remove logs please

**Comment by jamsea** [comment]
@doctyper Looks like you have to rebuild the project `docker-compose up --build` it's launching an old image.

File: README.md [hunk 12]
```diff
@@ -60,27 +60,51 @@ For first-time jspm users, you will need to configure it to use GitHub:
...
+```
+nameserver 127.0.0.1
+port 53535
+```
+
+Run:
+- `cd example`
+- `docker-compose up`
+- Open `https://www.example.dev:3000`
+
```
Comment (doctyper) [inline README.md]: Can you add a step linking to 'Accepting the development SSL certificate` below?

**Pull request merged**